pkg/scaler/prober.go

```diff
@@ -538,14 +592,18 @@ func (p *prober) scaleUp(ctx context.Context) error {
 	})
 }
 
-// Checks for a given resource considered for scale, if for the respecitve scale operations its dependent deployments are in desired state.
+// Checks for a given resource considered for scale, if for the respective scale operations its dependent deployments are in desired state.
 // If availableReplicas is not equal to desired then it fails the check and the scaling fo the parent is stopped
 func (p *prober) checkScaleRefDependsOn(ctx context.Context, prefix string, dependsOnScaleRefs []autoscalingapi.CrossVersionObjectReference, replicas int32, checkFn func(oReplicas, nReplicas int32) bool) bool {
+	// running this check immediately after scaling tends to fail as the parent resource might still be processing
+	// introduce a short delay to let the parent resource availability to reflect current state correctly
+	// TODO: We should replace this with a proper flag for DWD probe, this requires also to identify the currect default value.
+	time.Sleep(toDuration(pointer.Int32Ptr(10), 0))
```

Comment:
Not clear on what is the need to add 10s. Call to `checkScaleRefDependsOn` happens before the actual call to scale is made in function `retry(msg, p.getScalingFunc, gr, s, replicas)`

Reply:
In the current common way of handling scaleDown and scaleUp we observed that during scaleDown without any delay. The scale down of dependent resources is not handled in the same reconciliation step. 
for e.g. if `KCM` is scale down and immediately after `MCM` is tried as it is dependent on `KCM` it checks if the available replicas for `KCM` is 0. As this is done immediately after `KCM` is scaled down, we observed that for most of the time it's not the case as the cache is not updated. So it fails and then skips scaling down `MCM` and then is retried in the next cycle. 

Eventually everything will be scaled down but it delays down everything by (no. of scaleDependents) * 30s to finally scale everything down. Hence on a safer side a 10s delay ensures that we scale down everything by (no. of scaledDependents) * 10s.

Reply:
ok, but 10s delay is a magic number and there is no guarantee that either the parent dependency is scaled down and the cache is also refreshed. The correct way of doing this is to have a pipeline of stages for scale down and scale up. Each step in the pipeline waits for its dependency to scale down. Once that is done in a timeout amount of time then it also scales down child dependencies, else it skips to the next cycle.
As also pointed out earlier, there is no strict parent <-> child relationship betwee `KCM` and `MCM`. If the external probe fails and resultRun > failureThreshold then both `KCM` and `MCM` can be stopped together. This again easily possible using a pipeline model as there can be sequential pipeline steps of parallel. For now you can go with waiting for 10s but please note that this is not always going to give you the desired outcome.

Reply:
Totally in line with your concern / proposal. 
 This is only for the current approach and keeping in mind that we wish to refactor the flow, I believe we can address it there.

Reply:
ok.